### PR TITLE
Parser expand

### PR DESCRIPTION
**Pull request by jjsjann123**
1. Updated parser with expand/expand_as support. -> only fuse functional expand(_as) to avoid aliasing, handled the same way as with our other `alias_copy`.
2. updated `expand(_as)` in arith.cpp to allow rank broadcast on inner dimensions. (matching pytorch API).
3. python tests

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -1048,12 +1059,23 @@ TensorView* expand_as(TensorView* inp, TensorView* other) {
...
+  // broadcast inner on inp to match rank with other.
+  if (inp_domain.size() < other_domain.size()) {
+    const int num_bcast =
+        static_cast<int>(other_domain.size() - inp_domain.size());
+    std::vector<bool> inner_bcast_dims(other_domain.size(), false);
+    std::fill(
+        inner_bcast_dims.begin(), inner_bcast_dims.begin() + num_bcast, true);
+    inp = broadcast(inp, inner_bcast_dims);
+    inp_domain = TensorDomain::noReductions(inp->getMaybeRFactorDomain());
+  }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: Can you please factor this part out so that it can be used from both `expand` and `expand_as`?

**Review by naoyam — COMMENTED** [review]
Approving the C++ side.

**Review by naoyam — APPROVED** [review]

**Comment by rdspring1** [comment]
LGTM

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 2]
```diff
@@ -975,12 +975,23 @@ TensorView* expand(TensorView* inp, const std::vector<Val*>& expanded_sizes) {
...
       " but received ",
       expanded_sizes.size());
 
+  // broadcast inner on inp to match rank with other.
+  if (inp_domain.size() < expanded_sizes.size()) {
+    const int num_bcast =
+        static_cast<int>(expanded_sizes.size() - inp_domain.size());
+    std::vector<bool> inner_bcast_dims(expanded_sizes.size(), false);
+    std::fill(
+        inner_bcast_dims.begin(), inner_bcast_dims.begin() + num_bcast, true);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: This inserts broadcast axes to outside of inp_domain, i.e., left of the input domains. Is it what you intend to do?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: Yes it is. See note in PR (could use better English :stuck_out_tongue: ): `updated expand(_as) in arith.cpp to allow rank broadcast on inner dimensions. (matching pytorch API).`

```
In [1]: import torch

In [2]: x = torch.randn(5, 5)

In [3]: x.expand((2, 5, 5)).shape
Out[3]: torch.Size([2, 5, 5])

```

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: I can back this out into integration code if we think it's a crazy API tho.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: I don't see any problem, so no need to back it out.

**Pull request merged**